tmt/steps/provision/bootc.py

```diff
@@ -0,0 +1,242 @@
+    machine:
+
+    .. code-block:: shell
+
+        podman machine init --rootful --disk-size 200 --memory 8192 \
+        --cpus 8 -v /var/tmp/tmt:/var/tmp/tmt -v $HOME:$HOME
+    """
+
+    _data_class = BootcData
+    _guest_class = GuestTestcloud
+    _guest = None
+
+    def _get_id(self) -> str:
+        # FIXME: cast() - https://github.com/teemtee/tmt/issues/1372
+        parent = cast(tmt.steps.provision.Provision, self.parent)
+        assert parent.plan is not None
+        assert parent.plan.my_run is not None
+        assert parent.plan.my_run.unique_id is not None
+        return parent.plan.my_run.unique_id
+
+    def _expand_path(self, relative_path: str) -> str:
+        """ Expand the path to the full path relative to the current working dir """
+        if relative_path.startswith("/"):
+            return relative_path
+        return f"{os.getcwd()}/{relative_path}"
+
+    def _build_derived_image(self, base_image: str) -> str:
+        """ Build a "derived" container image from the base image with tmt dependencies added """
+        assert self.workdir is not None  # narrow type
+
+        self._logger.debug("Building modified container image with necessary tmt packages/config")
+        containerfile_template = '''
+            FROM {{ base_image }}
+
+            RUN \
+            dnf -y install cloud-init rsync && \
+            ln -s ../cloud-init.target /usr/lib/systemd/system/default.target.wants && \
+            rm /usr/local -rf && ln -sr /var/usrlocal /usr/local && mkdir -p /var/usrlocal/bin && \
+            dnf clean all
+        '''
+        containerfile_parsed = render_template(
+            containerfile_template,
+            base_image=base_image)
+        (self.workdir / 'Containerfile').write_text(containerfile_parsed)
+
+        image_tag = f'localhost/tmtmodified-{self._get_id()}'
+        tmt.utils.Command(
+            "podman", "build", f'{self.workdir}',
+            "-f", f'{self.workdir}/Containerfile',
+            "-t", image_tag
+            ).run(cwd=self.workdir, stream_output=True, logger=self._logger)
+
+        return image_tag
+
+    def _build_base_image(self, containerfile: str, workdir: str) -> str:
+        """ Build the "base" or user supplied container image """
+        image_tag = f'localhost/tmtbase-{self._get_id()}'
+        self._logger.debug("Building container image")
+        tmt.utils.Command(
+            "podman", "build", self._expand_path(workdir),
+            "-f", self._expand_path(containerfile),
+            "-t", image_tag
+            ).run(cwd=self.workdir, stream_output=True, logger=self._logger)
+        return image_tag
+
+    def _build_bootc_disk(self, containerimage: str, image_builder: str) -> None:
+        """ Build the bootc disk from a container image using bootc image builder """
+        self._logger.debug("Building bootc disk image")
+        tmt.utils.Command(
+            "podman", "run", "--rm", "--privileged",
```

Comment:
Whould this not run under root user for it to work? I get this error on my laptop when trying to run it:

```
❯ pwd
/var/home/mvadkert/git/github.com/teemtee/tmt/tests/provision/bootc/data
❯ tmt run -vvv -e TMT_BOOTC_CONTAINERFILE_RUNDIR=.
/var/home/mvadkert/.local/share/tmt/run-002
    warn: /containerfile_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/includes_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
    warn: /containerfile_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/needs_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
    warn: /image_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerimage': 'localhost/tmt-bootc-includes-deps', 'disk': 20} is not valid under any of the given schemas
    warn: /image_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerimage': 'localhost/tmt-bootc-needs-deps', 'disk': 20} is not valid under any of the given schemas
Found 4 plans.

/containerfile_includes_deps
    discover
        how: shell
        order: 50
        summary: 1 test selected
            /booted image
    provision
        queued provision.provision task #1: default-0
        
        provision.provision task #1: default-0
        how: bootc
        order: 50
        name: default-0
        how: bootc
        image: file:///var/home/mvadkert/.local/share/tmt/run-002/containerfile_includes_deps/provision/default-0/qcow2/disk.qcow2
        containerfile: ./includes_deps.containerfile
        add deps: False
        memory: 2048 MB
        disk: 20 GB
        fail: Command 'podman run --rm --privileged -v /var/lib/containers/storage:/var/lib/containers/storage --security-opt label=type:unconfined_t -v /var/home/mvadkert/.local/share/tmt/run-002/containerfile_includes_deps/provision/default-0:/output quay.io/centos-bootc/bootc-image-builder:latest build --type qcow2 --local localhost/tmtbase-1729527982' returned 1.
    finish
        warn: Nothing to finish, no guests provisioned.

plan failed

The exception was caused by 1 earlier exceptions

Cause number 1:

    provision step failed

    The exception was caused by 1 earlier exceptions

    Cause number 1:

        Command 'podman run --rm --privileged -v /var/lib/containers/storage:/var/lib/containers/storage --security-opt label=type:unconfined_t -v /var/home/mvadkert/.local/share/tmt/run-002/containerfile_includes_deps/provision/default-0:/output quay.io/centos-bootc/bootc-image-builder:latest build --type qcow2 --local localhost/tmtbase-1729527982' returned 1.

        stderr (8/8 lines)
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        Trying to pull quay.io/centos-bootc/bootc-image-builder:latest...
        Getting image source signatures
        Copying blob sha256:311336741819859e92064b81199a47397788d736d764f561a89cba9583600387
        Copying blob sha256:35056b112ca2a75be2e91b1076785d9712b1790aea49ebfd61f1799d090d9057
        Copying blob sha256:2edf720f10afc7137e2387bdb3ba800eec2602ac16da698ec19642a4d6fbc4dd
        Copying config sha256:767a10b2f63a72d479d9b102c2016790c63b045f6b9533c639fed605c299ee14
        Writing manifest to image destination
        2024/10/21 16:27:47 error: cannot validate the setup: this command must be run in rootful (not rootless) podman
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

per:

https://github.com/osbuild/bootc-image-builder?tab=readme-ov-file#building

And we will need to put it under the `--feeling-safe` option.

Personally I feel this might a bit limit the usability here, I am wondering if it would not make more sense to spin a VM and run this command in it, but yeah, that will make this a lot harder to do, but avoid the `root` requirement of this feature.

Reply:
We can do that later, any idea if this privileged root requirement will ever go away?

Reply:
currently, creating the bootc disk requires a loopback mount which is why root is required. I don't think that requirement will go away anytime soon, although it's a common pain point so it might happen sooner rather than later.

> Personally I feel this might a bit limit the usability here, I am wondering if it would not make more sense to spin a VM and run this command in it

A common pattern is to run podman-machine (a VM) to enable rootful podman. Then default all podman commands to use the machine, e.g.

```
podman machine init --rootful --disk-size 200 --memory 8192 --cpus 8 -v /var/tmp/tmt:/var/tmp/tmt -v $HOME:$HOME
export CONTAINER_CONNECTION=podman-machine-default-root
podman run ...
```

Reply:
this is documented in the PR here: https://github.com/teemtee/tmt/pull/3161/files#diff-f05fc41435a3de82fd67c56c9bb94d0b781ddaf4182fdf907b4d02ce3531476eR129

Reply:
@ckyrouac can you please incorporate `podman machine init` so we do not need to care about it, i.e for the time of building the image, it would be done automatically, not requiring any user action.

Reply:
@ckyrouac could we detect if running as root and have a default based on that? Seems like a good middle ground. And make an option to be able to force whatever if that is what the user wants?